### PR TITLE
DRAFT multiview_zero123

### PR DESCRIPTION
**Pull request by voletiv**
Adds ability to use multiple images i.e. multiple reference views of the same scene with Zero123.

Zero123 is conditioned on a single reference image and a relative pose. This PR adds the ability to use multiple reference images and their respective relative pose from the first reference image to generate a novel view using the same Zero123 model.

- It combines the noise prediction from each reference view using weights calculated using the relative pose of the novel view with respect to the reference views.
- Uses `--images_config` to take info about the multiple images. It retains the capability of simply using `--image`. Check the csv files in `config` for examples.
- The csv files use a left-handed spherical coordinate system to specify the poses.
- The csv files also include `zero123_weight` which is used to specify the weight of a reference view in novel view prediction  using Zero123. For example, if the `zero123_weight` of a reference image is 0, it will only be used for rgbd_loss, but not for zero123 loss.
- Also enables >1 batch size i.e. `num_images_per_batch`. By default, it is set to 1.
- Also introduces `test_freq` to generate mp4 videos at intermediate epochs as well (and not just the last)

### CODE REVIEW

**Comment by ashawkey** [comment]
This is really nice! As we discussed earlier, I just pushed a commit including lots of refactors, which causes some conflicts here.

**Comment by voletiv** [comment]
> This is really nice! As we discussed earlier, I just pushed a commit including lots of refactors, which causes some conflicts here.

Yeah I saw that, will resolve conflicts and push a commit!

**Comment by ashawkey** [comment]
Thanks! Basically the changes are:
* using a module dict to take care of multiple guidance models
* the `-azimuth` PR (https://github.com/ashawkey/stable-dreamfusion/pull/241) is moved into `zero123_utils.py` (https://github.com/ashawkey/stable-dreamfusion/blob/main/guidance/zero123_utils.py#L124)
* `preprocess_image.py` now recenter the object (and adjust border ratio) by default, which may need some modification for multi-view settings?
* some hyper-parameters that (seemingly) improve the quality

**Comment by voletiv** [comment]
@ashawkey Amazing, thanks for the summary! This was very helpful.
I've corrected the merge conflicts, please check!

**Comment by ashawkey** [comment]
Thanks! There's lot of changes so I'll take some time to check, I'll @ you once finished.

**Comment by claforte** [comment]
> The csv files use a left-handed spherical coordinate system to specify the poses.

I recommend we stick with right-handed coordinate system for consistency.. @ashawkey what do you think?

**Comment by claforte** [comment]
> I see you replaced the original large images with 512x512 smaller ones, but they are still in the commit history and makes git clone slow. It's better to totally delete them from the history using [this](https://stackoverflow.com/questions/2100907/how-to-remove-delete-a-large-file-from-commit-history-in-the-git-repository).

@ashawkey is that really required if you do a squash-and-merge? (I'd expect github to only keep the final versions of the files... but I'm not sure if that's how it works)

File: .gitignore [hunk 1]
```diff
@@ -18,6 +18,11 @@ TOKEN
 densegridencoder
 tets/256_tets.npz
 
+assets/
+launch.json
+settings.json
+
+aa*.png
```
Comment (claforte) [inline .gitignore:25]: I think all these should be removed... e.g. launch.json is already in there, below (`.vscode/launch.json`)

File: config/car.csv [hunk 2]
```diff
@@ -0,0 +1,5 @@
+zero123_weight, radius, theta, phi, image
```
Comment (claforte) [inline config/car.csv:1]: personally I prefer `polar` and `azimuth` instead of `theta` and `phi`. @ashawkey I'm not saying we should rename them everywhere, but since this csv file will be human-authored, we should stick with names that are easily understood and remembered...

File: data/anya_60.png
Comment (claforte) [inline data/anya_60.png:1]: why are we committing this one? it's a (bad) result of zero123...

File: data/car_back.jpg
Comment (claforte) [inline data/car_back.jpg:1]: @voletiv you might want to change or obscure the image's license plate... just in case. ;-)

File: guidance/zero123_utils.py [hunk 3]
```diff
@@ -53,35 +53,39 @@ def load_model_from_config(config, ckpt, device, vram_O=False, verbose=False):
...
+                 config='./pretrained/zero123/sd-objaverse-finetune-c_concat-256.yaml',
+                 ckpt='./pretrained/zero123/105000.ckpt', vram_O=False, t_range=[0.02, 0.98]):
         super().__init__()
 
-        # hardcoded
-        config = './pretrained/zero123/sd-objaverse-finetune-c_concat-256.yaml'
-        ckpt = './pretrained/zero123/105000.ckpt'
+        # # hardcoded
+        # config = './pretrained/zero123/sd-objaverse-finetune-c_concat-256.yaml'
+        # ckpt = './pretrained/zero123/105000.ckpt'
```
Comment (claforte) [inline guidance/zero123_utils.py:75]: IMHO these lines should simply be deleted.

File: guidance/zero123_utils.py [hunk 4]
```diff
@@ -115,23 +119,50 @@ def train_step(self, embeddings, pred_rgb, polar, azimuth, radius, guidance_scal
 
         t = torch.randint(self.min_step, self.max_step + 1, [1], dtype=torch.long, device=self.device)
 
+        # Set weights acc to closeness in phi
+        if len(embeddings['phis']) > 1:
+            phi_deltas = [abs(azimuth.item() + embeddings['phis'][0] - phi) for phi in embeddings['phis']]
```
Comment (claforte) [inline guidance/zero123_utils.py:124]: I think it would be better to calculate the weights according to the cosine of the orientation (and clamp it to [0,1]) instead of based on azimuth. Doing it on azimuth alone will not work, e.g. if we have a top, bottom and front views, that all have the same azimuth value of 0.

File: main.py [hunk 5]
```diff
@@ -74,6 +77,7 @@ def __call__ (self, parser, namespace, values, option_string = None):
     # try this if CUDA OOM
     parser.add_argument('--fp16', action='store_true', help="use float16 for training")
     parser.add_argument('--vram_O', action='store_true', help="optimization for low VRAM usage")
+    parser.add_argument('--batch_size_1', action='store_true', help="render only one image using NeRF")
```
Comment (claforte) [inline main.py:80]: I think it would be cleaner to use an `int` `num_images_per_batch` that defaults to 1. Then people with lots of VRAM could crank it up to increase performance and/or quality.

File: main.py [hunk 6]
```diff
@@ -95,6 +99,7 @@ def __call__ (self, parser, namespace, values, option_string = None):
     parser.add_argument('--default_theta', type=float, default=90, help="radius for the default view")
     parser.add_argument('--default_phi', type=float, default=0, help="radius for the default view")
     parser.add_argument('--default_fovy', type=float, default=20, help="fovy for the default view")
+    parser.add_argument('--default_img_w', type=float, default=1, help="zero123 weight for the default view")
```
Comment (claforte) [inline main.py:102]: Not sure if that's really required. If there's just 1 reference view, why would you ever set this to a value other than 1?

File: main.py [hunk 7]
```diff
@@ -103,6 +108,8 @@ def __call__ (self, parser, namespace, values, option_string = None):
     parser.add_argument('--angle_front', type=float, default=60, help="[0, angle_front] is the front region, [180, 180+angle_front] the back region, otherwise the side region.")
     parser.add_argument('--t_range', type=float, nargs='*', default=[0.02, 0.98], help="stable diffusion time steps range")
 
+    parser.add_argument('--test_freq', type=int, default=None, help="Test model every n iterations")
```
Comment (claforte) [inline main.py:111]: I'd clarify it, e.g. replace "Test model every n iterations" with "test the model every n iteration, by recording a turntable video. (By default, this is only done after the final epoch)"

File: main.py [hunk 8]
```diff
@@ -135,6 +142,15 @@ def __call__ (self, parser, namespace, values, option_string = None):
     parser.add_argument('--light_phi', type=float, default=0, help="default GUI light direction in [0, 360), azimuth")
     parser.add_argument('--max_spp', type=int, default=1, help="GUI rendering max sample per pixel")
 
+    parser.add_argument('--zero123_config', type=str, default='./pretrained/zero123/sd-objaverse-finetune-c_concat-256.yaml', help="config file for zero123")
+    parser.add_argument('--zero123_ckpt', type=str, default='./pretrained/zero123/105000.ckpt', help="ckptfor zero123")
```
Comment (claforte) [inline main.py:146]: "ckptfor zero123" -> ""ckpt for zero123"

File: main.py [hunk 9]
```diff
@@ -135,6 +142,15 @@ def __call__ (self, parser, namespace, values, option_string = None):
...
+
+    parser.add_argument('--dataset_size_train', type=int, default=100, help="Length of train dataset")
+    parser.add_argument('--dataset_size_valid', type=int, default=8, help="Length of train dataset")
```
Comment (claforte) [inline main.py:149]: you meant valid/test, not 3 times "train dataset". ;-)

Comment (claforte) [reply main.py:149]: Anyway... I'm not sure what this is used for?

File: main.py [hunk 10]
```diff
@@ -135,6 +142,15 @@ def __call__ (self, parser, namespace, values, option_string = None):
...
+    parser.add_argument('--dataset_size_test', type=int, default=100, help="Length of train dataset")
+
+    parser.add_argument('--debug', action='store_true', help="debug options")
```
Comment (claforte) [inline main.py:152]: "debug" and "debug options" aren't sufficiently clear... what is this used for?

**Comment by ashawkey** [comment]
> > I see you replaced the original large images with 512x512 smaller ones, but they are still in the commit history and makes git clone slow. It's better to totally delete them from the history using [this](https://stackoverflow.com/questions/2100907/how-to-remove-delete-a-large-file-from-commit-history-in-the-git-repository).
> 
> @ashawkey is that really required if you do a squash-and-merge? (I'd expect github to only keep the final versions of the files... but I'm not sure if that's how it works)

You are right! I did't know that.

File: config/car.csv (hunk 2 above)
Comment (ashawkey) [reply config/car.csv:1]: I agree, `polar/azimuth` may be better.

Comment (voletiv) [reply config/car.csv:1]: Replaced everywhere :)

**Comment by voletiv** [comment]
@ashawkey @claforte Updated the code according to your reviews and merged conflicts, please check!

**Comment by ashawkey** [comment]
@voletiv Thanks! My last concern is about the performance of multi-view setting, but that can be improved later. If @claforte has no problem, I'll merge it.

**Comment by voletiv** [comment]
@ashawkey Haha yeah, I think the source of that problem is the radius,polar,azimuth of the back image not being accurate. I have other experiments with other values where the results are even worse with ghost parts floating in the air. The one that you sent is actually much better!

**Comment by claforte** [comment]
Sorry, I just caught up right now (I got side-tracked). I'll do the squash-and-merge now. Thank you!

**Comment by taeyeopl** [comment]
Hi, @voletiv @claforte! Thanks for the multiview support! try to use the multiview condition but I fail to get the results. I guess the problem comes from the input object size of the front/back and left/right sides having different foreground sizes. I try to make it the same size with --dont_recenter but it is still not works. Can you share some successful multiview examples of car codes? 

Commands:
# Preprocess
python preprocess_image.py data/car_front.jpg --dont_recenter
python preprocess_image.py data/car_back.jpg --dont_recenter
python preprocess_image.py data/car_left.jpg --dont_recenter
python preprocess_image.py data/car_right.jpg --dont_recenter
# Training Initial Results
python main.py -O --image_config config/car.csv --workspace trial_image/car_four_no_recenter --iters 5000
# Training Refine Results
python main.py -O --image_config config/car.csv --workspace trial_image/car_four_no_recenter_refine --iters 10000 --dmtet --init_with trial_image/car_four_no_recenter/checkpoints/df.pth


https://github.com/ashawkey/stable-dreamfusion/assets/50736858/2d8d3485-3ee7-4b66-a27d-a1bc69f79f92


https://github.com/ashawkey/stable-dreamfusion/assets/50736858/6e66c696-17cf-4fac-9100-10fee09bbe0d

**Comment by 523997931** [comment]
> Hi, @voletiv @claforte! Thanks for the multiview support! I try to use the multiview condition but I fail to get good results. I guess the problem comes from the input object size of the front/back and left/right sides having different object sizes. I try to make it the same size with --dont_recenter but it is still not works. Can you share some successful multiview examples of car codes? Any advice is also welcome.
> 
> Commands:
> 
> # Preprocess
> python preprocess_image.py data/car_front.jpg --dont_recenter python preprocess_image.py data/car_back.jpg --dont_recenter python preprocess_image.py data/car_left.jpg --dont_recenter python preprocess_image.py data/car_right.jpg --dont_recenter
> 
> # Training Initial Results
> python main.py -O --image_config config/car.csv --workspace trial_image/car_four_no_recenter --iters 5000
> 
> # Training Refine Results
> python main.py -O --image_config config/car.csv --workspace trial_image/car_four_no_recenter_refine --iters 10000 --dmtet --init_with trial_image/car_four_no_recenter/checkpoints/df.pth
> 
>  initial_rgb.mp4 
>  refine_rgb.mp4

hello，

**Comment by 523997931** [comment]
> Hi, @voletiv @claforte! Thanks for the multiview support! I try to use the multiview condition but I fail to get good results. I guess the problem comes from the input object size of the front/back and left/right sides having different object sizes. I try to make it the same size with --dont_recenter but it is still not works. Can you share some successful multiview examples of car codes? Any advice is also welcome.
> 
> Commands:
> 
> # Preprocess
> python preprocess_image.py data/car_front.jpg --dont_recenter python preprocess_image.py data/car_back.jpg --dont_recenter python preprocess_image.py data/car_left.jpg --dont_recenter python preprocess_image.py data/car_right.jpg --dont_recenter
> 
> # Training Initial Results
> python main.py -O --image_config config/car.csv --workspace trial_image/car_four_no_recenter --iters 5000
> 
> # Training Refine Results
> python main.py -O --image_config config/car.csv --workspace trial_image/car_four_no_recenter_refine --iters 10000 --dmtet --init_with trial_image/car_four_no_recenter/checkpoints/df.pth
> 
>  initial_rgb.mp4 
>  refine_rgb.mp4

hello, 
this days, i also try to use multi view car picture to construction 3d model , and I met the same problems with you, the shape is so weird on the front wheel and top of car
do you find any method to deal with it?
